subsys/bluetooth/host/l2cap.c

```diff
@@ -938,14 +938,16 @@ struct net_buf *l2cap_data_pull(struct bt_conn *conn,
 		__maybe_unused struct net_buf *b = k_fifo_get(&lechan->tx_queue, K_NO_WAIT);
 
 		__ASSERT_NO_MSG(b == pdu);
+	}
 
-		if (L2CAP_LE_CID_IS_DYN(lechan->tx.cid)) {
-			LOG_DBG("adding `sdu_sent` callback");
-			/* No user callbacks for SDUs */
-			make_closure(pdu->user_data,
-				     l2cap_chan_sdu_sent,
-				     UINT_TO_POINTER(lechan->tx.cid));
-		}
+	if (L2CAP_LE_CID_IS_DYN(lechan->tx.cid)) {
+		bool sdu_end = last_frag && last_seg;
+
+		LOG_DBG("adding `sdu_sent` callback");
+		/* No user callbacks for SDUs */
+		make_closure(pdu->user_data,
+			     sdu_end ? l2cap_chan_sdu_sent : NULL,
+			     sdu_end ? UINT_TO_POINTER(lechan->tx.cid) : NULL);
```

Comment:
Since the addition of the `l2cap_chan_sdu_sent` callback is now conditional, maybe the log statement should reflect that. As the commit is now, "adding `sdu_sent` callback" may be logged even when the callback is set to NULL.

Reply:
That's true. I had two log messages before, but moved around the ifs and forgot this log message.